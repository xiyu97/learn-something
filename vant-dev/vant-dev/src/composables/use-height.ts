import { useRect } from '@vant/use';
import { Ref, ref, onMounted, nextTick } from 'vue';

export const useHeight = (element: Element | Ref<Element | undefined>) => {
  const height = ref<number>();

  onMounted(() => {
    nextTick(() => {
      height.value = useRect(element).height;
    });
  });

  return height;
};
