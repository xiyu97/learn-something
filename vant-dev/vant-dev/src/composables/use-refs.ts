import { ref, Ref, onBeforeUpdate } from 'vue';

export function useRefs<T = Element>() {
  const refs = ref([]) as Ref<T[]>;

  onBeforeUpdate(() => {
    refs.value = [];
  });

  const setRefs = (index: number) => (el: unknown) => {
    refs.value[index] = el as T;
  };

  return [refs, setRefs] as const;
}
