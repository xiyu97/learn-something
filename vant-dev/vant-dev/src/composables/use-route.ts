/**
 * Vue Router support
 */
import {
  PropType,
  ExtractPropTypes,
  getCurrentInstance,
  ComponentPublicInstance,
} from 'vue';
import type { RouteLocationRaw } from 'vue-router';

export const routeProps = {
  to: [String, Object] as PropType<RouteLocationRaw>,
  url: String,
  replace: Boolean,
};

export type RouteProps = ExtractPropTypes<typeof routeProps>;

export function route(vm: ComponentPublicInstance<RouteProps>) {
  const router = vm.$router;
  const { to, url, replace } = vm;

  if (to && router) {
    router[replace ? 'replace' : 'push'](to);
  } else if (url) {
    replace ? location.replace(url) : (location.href = url);
  }
}

export function useRoute() {
  const vm = getCurrentInstance()!.proxy as ComponentPublicInstance<RouteProps>;
  return () => route(vm);
}
