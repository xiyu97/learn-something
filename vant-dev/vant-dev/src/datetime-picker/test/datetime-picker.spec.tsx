import { DatetimePicker } from '..';
import { mount, later } from '../../../test';
import { reactive } from 'vue';
import { useExpose } from '../../composables/use-expose';

test('confirm & cancel event', () => {
  const onConfirm = jest.fn();
  const onCancel = jest.fn();

  const wrapper = mount(DatetimePicker, {
    props: {
      onConfirm,
      onCancel,
    },
  });

  wrapper.find('.van-picker__confirm').trigger('click');
  expect(onConfirm).toHaveBeenCalledTimes(1);

  wrapper.find('.van-picker__cancel').trigger('click');
  expect(onCancel).toHaveBeenCalledTimes(1);
});

test('time type', () => {
  const wrapper = mount(DatetimePicker, {
    props: {
      type: 'time',
      minHour: 22,
      minMinute: 58,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('getPicker method', () => {
  const wrapper = mount(DatetimePicker);

  expect(wrapper.vm.getPicker()).toBeTruthy();
});

test('should render title slot correctly', () => {
  const wrapper = mount(DatetimePicker, {
    props: {
      showToolbar: true,
    },
    slots: {
      title: () => 'Custom title',
    },
  });

  expect(wrapper.find('.van-picker__toolbar').html()).toMatchSnapshot();
});

test('should emit value correctly when dynamic change min-date', async () => {
  const defaultValue = new Date(2020, 10, 2, 10, 30);
  const wrapper = mount({
    emits: ['confirm'],
    setup(_, { emit }) {
      const state = reactive({
        date: defaultValue,
        minDate: new Date(2010, 0, 1, 10, 30),
      });

      const onChange = () => {
        state.minDate = state.date;
      };

      useExpose({
        onChange,
      });

      return () => (
        <DatetimePicker
          v-model={state.date}
          minDate={state.minDate}
          onConfirm={(value: Date) => emit('confirm', value)}
        />
      );
    },
  });

  await later();
  (wrapper.vm as Record<string, any>).onChange();
  await later();
  wrapper.find('.van-picker__confirm').trigger('click');
  expect(wrapper.emitted<[Date]>('confirm')![0][0]).toEqual(defaultValue);
});
