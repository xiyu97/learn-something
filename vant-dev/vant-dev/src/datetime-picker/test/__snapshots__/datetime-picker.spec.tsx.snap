// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render title slot correctly 1`] = `
<div class="van-picker__toolbar">
  <button type="button"
          class="van-picker__cancel"
  >
    Cancel
  </button>
  Custom title
  <button type="button"
          class="van-picker__confirm"
  >
    Confirm
  </button>
</div>
`;

exports[`time type 1`] = `
<div class="van-picker van-datetime-picker">
  <div class="van-picker__toolbar">
    <button type="button"
            class="van-picker__cancel"
    >
      Cancel
    </button>
    <button type="button"
            class="van-picker__confirm"
    >
      Confirm
    </button>
  </div>
  <div class="van-picker__columns"
       style="height: 264px;"
  >
    <div class="van-picker-column">
      <ul style="transform: translate3d(0, 110px, 0); transition-duration: 0ms; transition-property: none;"
          class="van-picker-column__wrapper"
      >
        <li role="button"
            style="height: 44px;"
            tabindex="0"
            class="van-picker-column__item van-picker-column__item--selected"
        >
          <div class="van-ellipsis">
            22
          </div>
        </li>
        <li role="button"
            style="height: 44px;"
            tabindex="0"
            class="van-picker-column__item"
        >
          <div class="van-ellipsis">
            23
          </div>
        </li>
      </ul>
    </div>
    <div class="van-picker-column">
      <ul style="transform: translate3d(0, 110px, 0); transition-duration: 0ms; transition-property: none;"
          class="van-picker-column__wrapper"
      >
        <li role="button"
            style="height: 44px;"
            tabindex="0"
            class="van-picker-column__item van-picker-column__item--selected"
        >
          <div class="van-ellipsis">
            58
          </div>
        </li>
        <li role="button"
            style="height: 44px;"
            tabindex="0"
            class="van-picker-column__item"
        >
          <div class="van-ellipsis">
            59
          </div>
        </li>
      </ul>
    </div>
    <div class="van-picker__mask"
         style="background-size: 100% 110px;"
    >
    </div>
    <div class="van-hairline-unset--top-bottom van-picker__frame"
         style="height: 44px;"
    >
    </div>
  </div>
</div>
`;
